### PR TITLE
refactor rpc calls

### PR DESCRIPTION
**Pull request by Adam-Goldman7**
reolves #59 

I'm not sure if at this point there's sense into making the calls component generic, i.e.
passing it the actions instead of defining them directly inside

what do u think?

PS. will correct linting later

### CODE REVIEW

File: client/provider/dom-provider.js [hunk 1]
```diff
@@ -1,3 +1,30 @@
+
+import React from 'react';
+import { isArray, isPlainObject } from 'lodash';
+
 export function hasScrollbar (el) {
   return el.clientHeight < el.scrollHeight;
 }
+
+export function formatRenderedResponse (res) {
```
Comment (tomusdrw) [inline client/provider/dom-provider.js]: Why not putting that to separate component instead? (IMHO `provider` is starting to look like an eufemism for `util` :))

Comment (Adam-Goldman7) [inline client/provider/dom-provider.js]: formatRenderedResponse -> Response component

about dom-provider -> where would u put it?
`utils/dom-util.js`?

Comment (tomusdrw) [inline client/provider/dom-provider.js]: Why not a separate component?
`components/Json`? or `components/RpcJsonResponse`

Comment (Adam-Goldman7) [inline client/provider/dom-provider.js]: guess I was unclear 
I've moved it to a Response component

I'm asking about the hasScroolbar function
where would u place that

**Comment by Adam-Goldman7** [comment]
resolves #95

File: client/components/CallsToolbar/index.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+export default class CallsToolbar extends Component {
+
+  render () {
+    const {call, callEl, containerEl} = this.props;
+    if (!call) {
+      return null;
+    }
+
+    const wrapStyle = {top: callEl.offsetTop - 22 - containerEl.scrollTop};
```
Comment (tomusdrw) [inline client/components/CallsToolbar/index.js]: Let's start to put `magical numbers` into some constants with meaningful names.

File: client/components/RpcCalls/index.js [hunk 3]
```diff
@@ -29,21 +22,20 @@ export default class RpcCalls extends Component {
...
-        {...this._test('prev-calls-remove')}
-        title='Clear RPC calls history'
-        onClick={::this.props.actions.resetRpcPrevCalls}
-        className={styles.removeIcon}
-        >
-        <i className='icon-trash'></i>
-      </a>
-    );
+      if (this.state.jsonMode) {
+        ::this.setJsonEditorValue();
```
Comment (tomusdrw) [inline client/components/RpcCalls/index.js]: `::` no needed

File: client/components/RpcCalls/index.js [hunk 4]
```diff
@@ -29,21 +22,20 @@ export default class RpcCalls extends Component {
     this.state = {};
   }
 
-  renderClear () {
-    if (!this.props.rpc.prevCalls.length) {
-      return;
-    }
+  componentWillReceiveProps (nextProps) {
```
Comment (tomusdrw) [inline client/components/RpcCalls/index.js]: `const {selectedMethod} = nextProps.rpc;`

File: client/components/RpcCalls/index.js [hunk 5]
```diff
@@ -155,7 +89,8 @@ export default class RpcCalls extends Component {
         {this.renderFormEditor()}
         <button
           {...this._test('fireRpc')}
-          className={`dapp-block-button ${styles.button}`}
+          className={`dapp-block-button`}
+          disabled={this.state.jsonEditorError}
           onClick={() => ::this.onRpcFire() }
```
Comment (tomusdrw) [inline client/components/RpcCalls/index.js]: Without `::` or `onClick={::this.onRpcFire}`

File: client/components/RpcCalls/index.js [hunk 6]
```diff
@@ -237,86 +142,30 @@ export default class RpcCalls extends Component {
...
-    let method = _.find(rpcMethods, {name: call.name});
-    this.props.actions.selectRpcMethod(method);
-
-    // and set parameter values
-    method.params.map((param, idx) => {
-      this.setState({
-        [`params_${param}`]: call.params[idx]
-      });
+  onCustomRpcFire () {
+    const {jsonEditorParsedValue} = this.state;
```
Comment (tomusdrw) [inline client/components/RpcCalls/index.js]: How about 
```javascript
const {method, params} = this.state.jsonEditorParsedValue;
this.props.actions.fireRpc({method, params});
```

File: client/components/RpcCalls/index.js [hunk 7]
```diff
@@ -345,31 +194,14 @@ export default class RpcCalls extends Component {
...
   setJsonEditorValue () {
     const {selectedMethod} = this.props.rpc;
-    const method = {
-      name: selectedMethod.name,
-      params: selectedMethod.params.map(p => this.state[`params_${p}`]),
-      inputFormatters: selectedMethod.inputFormatters,
-      outputFormatter: selectedMethod.outputFormatter
+    const json = {
+      method: selectedMethod.name,
+      params: selectedMethod.params.map(p => this.state[`params_${p}`])
```
Comment (tomusdrw) [inline client/components/RpcCalls/index.js]: Could you extract `this.state['params_${p}']` to some function (e.g. `this.paramValue(p)`)?

File: client/util/error.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+export function hasErrors (xs) {
+  if (!xs) {
+    return;
+  }
+  return !!xs.find(isError);
+}
+
+export function isError (x) {
+  return Object.prototype.toString.call(x) === '[object Error]';
```
Comment (tomusdrw) [inline client/util/error.js]: `x instanceof Error` doesn't work?

File: test/e2e/status-page.js [hunk 9]
```diff
@@ -23,7 +23,7 @@ module.exports = {
   'Assert mining settings' (client) {
     [
       { selector: 'author', name: 'eth_coinbase' },
-      { selector: 'extra-data', name: 'ethcore_extraData' },
+      // { selector: 'extra-data', name: 'ethcore_extraData' },
```
Comment (tomusdrw) [inline test/e2e/status-page.js]: Thanks :)

File: client/components/AutoComplete/index.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        openOnFocus={this.openOnFocus()}
+        filter={this.filter()}
+        {...this.props}
+      />
+    );
+  }
+
+  openOnFocus () {
+    const { openOnFocus } = this.props;
+    if (typeof openOnFocus === 'undefined') {
```
Comment (tomusdrw) [inline client/components/AutoComplete/index.js]: You can use `defaultProps` for that.

File: client/components/AutoComplete/index.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+      return true;
+    }
+
+    return openOnFocus;
+  }
+
+  filter () {
+    return this.props.filter || this.defaultFilter();
```
Comment (tomusdrw) [inline client/components/AutoComplete/index.js]: `defaultProps`

**Comment by tomusdrw** [comment]
It's pretty substantial, so let's focus on fixing the critical stuff and all other refactorings can be done in separate PRs.

File: client/components/RpcCalls/index.js (hunk 5 above)
Comment (Adam-Goldman7) [inline client/components/RpcCalls/index.js]: right.
leftovers from refactoring

File: client/components/AutoComplete/index.js [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+  render () {
+    return (
+      <AutoComplete {...this.props} />
+    );
+  }
+
+}
+
+WrappedAutoComplete.propTypes = {
```
Comment (tomusdrw) [inline client/components/AutoComplete/index.js]: I think we can also use syntax from `stage-0`, might be worth checking:
```javascript
class WrappedAutoComplete {
 static propTypes = {...};
 static defaultProps = {...};

}
```

Comment (Adam-Goldman7) [inline client/components/AutoComplete/index.js]: AFAIK only methods can b defined like that in JS classes

Comment (tomusdrw) [inline client/components/AutoComplete/index.js]: https://github.com/jeffmo/es-class-fields-and-static-properties

**Pull request merged**